### PR TITLE
Gatsby V4 support

### PR DESCRIPTION
**Pull request by DanielSLew**
PR for updating plugin for Gatsby V4 compatibility.

We need to initialize the error map as well as any global state in `onPluginInit`. We also need a check that keeps the old implementation to ensure backwards compatibility with v3.

### CODE REVIEW

File: src/gatsby-node.ts [hunk 1]
```diff
@@ -56,17 +53,21 @@ const defaultConfig = {
 
 const stateCache: {[key: string]: any} = {}
 
-export const onPreInit: GatsbyNode['onPreInit'] = async ({reporter}: ParentSpanPluginArgs) => {
-  // Old versions of Gatsby does not have this method
-  if (reporter.setErrorMap) {
-    reporter.setErrorMap(ERROR_MAP)
-  }
-}
+// export const onPreInit: GatsbyNode['onPreInit'] = async ({reporter}: ParentSpanPluginArgs) => {
```
Comment (DanielSLew) [inline src/gatsby-node.ts:56]: Do we still need this here in order to not break V3 support?

File: src/gatsby-node.ts [hunk 2]
```diff
@@ -56,17 +53,21 @@ const defaultConfig = {
...
+//   // Old versions of Gatsby does not have this method
+//   if (reporter.setErrorMap) {
+//     reporter.setErrorMap(ERROR_MAP)
+//   }
+// }
 
-export const onPreBootstrap: GatsbyNode['onPreBootstrap'] = async (args, pluginOptions?) => {
+export const unstable_onPluginInit: GatsbyNode['onPreBootstrap'] = async (args, pluginOptions?) => {
```
Comment (DanielSLew) [inline src/gatsby-node.ts:63]: Do we need both `onPluginInit` and `onPreBootstrap` to ensure V3 and V4 support?

**Comment by hdoro** [comment]
Tests are passing and I've tested it locally with Gatsby v3 & v4 - looking good! Queries, real-time time previews and build step (local & in Gatsby Cloud) working fine. I'm worried about complex sites with type conflicts and advanced schema customizations, though.

@rexxars what other measures & tests should we do/run?

**Review by hdoro — APPROVED** [review]

**Comment by rexxars** [comment]
The only thing I'm worried about here is the mention of workers, and how that affects the `stateMap` that is initialized. Will that work across the lifecycle, or refer to different state maps in different workers/main process?

I'm happy to hear it appears to be working on both v3 and v4 - that does seem to indicate that the state map persists. I'm just not sure if this holds true in all environments - eg does it do different things in gatsby cloud vs local, or is node workers used to emulate the same behavior locally?

**Comment by DanielSLew** [comment]
> The only thing I'm worried about here is the mention of workers, and how that affects the `stateMap` that is initialized. Will that work across the lifecycle, or refer to different state maps in different workers/main process?
> 
> I'm happy to hear it appears to be working on both v3 and v4 - that does seem to indicate that the state map persists. I'm just not sure if this holds true in all environments - eg does it do different things in gatsby cloud vs local, or is node workers used to emulate the same behavior locally?

@rexxars In order to access the same state across all the workers we can use the cache. But if you want to just initialize state across the workers (ie like an ERROR_MAP or any global state) we can do this in `onPluginInit`, this new lifecycle method will run in each worker as it is initialized. Which is why we needed to move the initializing plugin logic into the `onPluginInit` lifecycle. Each worker will now have that initial plugin state.

The worker processes are created in the Gatsby framework itself, it has nothing to do with Gatsby Cloud, so the behaviour should be the same locally and on Gatsby Cloud

**Review by rexxars — APPROVED** [review]
Looks good to me :)

**Pull request merged**